Code/Tools/AssetProcessor/native/tests/assetmanager/JobDependencySubIdTests.cpp

```diff
@@ -0,0 +1,258 @@
+        ASSERT_TRUE(m_stateData->SetSource(source2));
+
+        JobDatabaseEntry job1{
+            source1.m_sourceID, "Mock Job", 1234, "pc", m_builderInfoHandler.m_builderDesc.m_busId, AzToolsFramework::AssetSystem::JobStatus::Completed, 999
+        };
+
+        ASSERT_TRUE(m_stateData->SetJob(job1));
+
+        ProductDatabaseEntry product1{
+            job1.m_jobID, 0, "product.txt", m_assetType, AZ::Uuid::CreateName("product.txt"),
+            hashA
+        };
+        ProductDatabaseEntry product2{ job1.m_jobID,
+                                       777,
+                                       "product777.txt",
+                                       m_assetType,
+                                       AZ::Uuid::CreateName("product777.txt"),
+                                       hashB };
+
+        ASSERT_TRUE(m_stateData->SetProduct(product1));
+        ASSERT_TRUE(m_stateData->SetProduct(product2));
+
+        SourceFileDependencyEntry dependency1{ AZ::Uuid::CreateRandom(),
+                                               source2.m_sourceName.c_str(),
+                                               source1.m_sourceName.c_str(),
+                                               SourceFileDependencyEntry::DEP_JobToJob,
+                                               0,
+                                                useSubId ? AZStd::to_string(product2.m_subID).c_str() : "" };
+
+        ASSERT_TRUE(m_stateData->SetSourceFileDependency(dependency1));
+    }
+
+    void JobDependencySubIdTest::RunTest(bool firstProductChanged, bool secondProductChanged)
+    {
+        AZ::IO::Path cacheDir(m_tempDir.GetDirectory());
+        cacheDir /= "Cache";
+
+        AZStd::string productPath = (cacheDir / "product.txt").AsPosix().c_str();
+        AZStd::string product2Path = (cacheDir / "product777.txt").AsPosix().c_str();
+
+        UnitTestUtils::CreateDummyFile(productPath.c_str(), "unit test file");
+        UnitTestUtils::CreateDummyFile(product2Path.c_str(), "unit test file");
+
+        AZ::u64 hashA = firstProductChanged ? 0 : AssetUtilities::GetFileHash(productPath.c_str());
+        AZ::u64 hashB = secondProductChanged ? 0 : AssetUtilities::GetFileHash(product2Path.c_str());
+
+        CreateTestData(hashA, hashB, true);
+
+        QMetaObject::invokeMethod(
+            m_assetProcessorManager.get(), "AssessModifiedFile", Qt::QueuedConnection, Q_ARG(QString, m_parentFile.c_str()));
+        QCoreApplication::processEvents();
+
+        m_assetProcessorManager->CheckActiveFiles(1);
+
+        // AssessModifiedFile is going to set up a OneShotTimer with a 1ms delay on it.  We have to wait a short time for that timer to
+        // elapse before we can process that event. If we use the alternative processEvents that loops for X milliseconds we could
+        // accidentally process too many events.
+        AZStd::this_thread::sleep_for(AZStd::chrono::milliseconds(10));
+        QCoreApplication::processEvents();
```

Comment:
I probably wrote this at some point somewhere but now that I look at it makes me really worried.

I'm not sure what the alternative is though... and won't block this review for fixing it.

Reply:
Are you referring to this actual piece of code here, or the code its referring to?  This is a new unit test I wrote, I deliberately made it to be different from the usual tests where we just loop processEvents for up to 5s because I was trying to establish a pattern of invoking AP in more discreet steps so I could verify each action along the way.  I had to put the sleep in though to get past that OneShotTimer.  I couldn't think of any other way to make the timer fire other than waiting sufficiently long.